Documentation tidyup/updates

bsian03 opened the pull request:
Continuation of #957 (or rather another attempt)
<details>
<summary>See here for old stuff</summary>
I haven't done JSDoc this time round, however have done the structure changes:

From inside the namespace, the order is as follows)
* Constants - Constants and VERSION
* Types - grouped together, alphabetical order of group name and in each group (i haven't included group name)
* Interfaces, also as above (Exception being Events and Constants, of which are at the bottom of their respective group)
* Classes, in alphabetical order

Each category has 2 newline separations, with each group, or each class with a single newline separation

I'll also fill in missing docs for JSDoc

Invite.guild also now gets the relevant guild
Invite.channel now gets the relevant guild channel if the guild is cached. if group channel or not cached this will still be an object

(note for abal, please don't merge this with the other PRs since there's bound to be conflicts once the other PRs get merged)

</details>

## Documentation/codebase layout changes
- Everything is now in alphabetical order, this includes all docs (including TypeScript) and methods/props.
- Added a linter to enforce the above (classes only)
- Added missing and corrected incorrect docs
- Consistency in JSDoc types
- Removed superclass props from JSDoc
- Invite#guild and Invite#channel will now fetch their respective instances. If Invite#guild is not cached, this will still be an object. If Invite#guild is not cached or Invite#channel is a GroupChannel, this will also still be an object. (this should really be in its own PR but too late)
- Deprecate supplying `userID` parameter for messageReactionAdd and messageReactionRemove for DMs (Anything other than "@me" will error)
- Update to TypeScript v4!

### [TypeScript v4](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/)
- Updated typescript eslint parser and plugin
- Slight change to types to respect TypeScript v4 rules
- Add deprecation flag to deprecated properties/getters/methods! (This also works for overloads)

Warning may be shown in the console of the CI, this can be disregarded, [see here](https://github.com/typescript-eslint/typescript-eslint/issues/2260#issuecomment-677853883)

Resolves #1003
Resolves #1004

bsian03 committed:
Partial of previously open PR

bsian03 committed:
Selfbot note for oauth app ID param

bsian03 committed:
reorder

bsian03 committed:
Typing diff for previous PR

bsian03 committed:
Add missing OPCodes for Voice

bsian03 committed:
Optional auth param

bsian03 committed:
Fix docs order

bsian03 committed:
Reoder everything, remove some duplicates

bsian03 committed:
Lint

jtsshieh commented on index.d.ts:
```diff
@@ -8,834 +8,369 @@ declare function Eris(token: string, options?: Eris.ClientOptions): Eris.Client;
+    permissionOverwrites: Collection<PermissionOverwrite>;
+    rateLimitPerUser?: number;
+    position: string;
+    topic?: string;
```
Discussed in #js_eris, topic can be null as well

bsian03 replied:
#981

jtsshieh replied:
okay

bsian03 committed:
Doc updates/changes

bsian03 committed:
Doc update

bsian03 committed:
Order

bsian03 committed:
Merge branch 'dev' of https://github.com/abalabahaha/eris into typings

bsian03 committed:
Some corrections

bsian03 committed:
Lint

bsian03 committed:
Don't break guild instantiation

bsian03 committed:
Add missing space

bsian03 committed:
Correct documentation

bsian03 committed:
Update event documentation

bsian03 committed:
Add missing Message props

bsian03 committed:
Correct webhooksUpdate doc

bsian03 committed:
Strict type PermissionOverwrite#type

bsian03 committed:
Merge branch 'dev' of https://github.com/abalabahaha/eris into typings

bsian03 committed:
Tidy some stuff

bsian03 committed:
Correct doc for EmojiOptions

bsian03 committed:
Correct some stuff

bsian03 committed:
Optional Message#application

bsian03 committed:
Correct docs

jtsshieh committed:
Alphabetize jsdoc comments and methods in classes (#2)

* Alphabetize oldPresence, oldState, oldMessage, oldRole, and emoji

also added oldMessage.pinned to typings. idk if it is supposed to be null or undefined though.

* Alphabetize oldCall and and relationship in friendSuggestionCreate

* Alphabetize props for Call, CategoryChannel, and Channel

* Alphabetize Guild

* Alphabetize GuildAuditLogEntry

* Alphabetize GuildIntegration

* Alphabetize GuildPreview

* Alphabetize Invite

* Alphabetize Member

* Alphabetize Message

* Fix PermissionOverwrite typing

Please correct if wrong, i don't write typings usually

* Alphabetize Role

* Alphabetize User

* Alphabetize VoiceState

* Alphabetize GuildChannel and fix typings for GuildChannel

* Alphabetize NewsChannel

* Alphabetize PrivateChannel

* Alphabetize StoreChannel

* Alphabetize PrivateChannel method args

* Alphabetize TextChannel

* Alphabetize VoiceChannel

please check if the typings are correct here

* Alphabetize Client

* Put Guild getters at the top

* Add messageReference and more client options

* Add semicolons

* Alphabetize Client methods

* Add bsian's commits

* Add compute prune count

* Alphabetize oldGuild and oldChannel

* Add back in createdAt

* Add new docs

* Alphabetize GroupChannel

* Fix conflicts in Guild

* Move permissionOf to correct location

* Remove extra space

Co-authored-by: bsian03 <chharry321@gmail.com>

jtsshieh committed:
Refer user to superclass props (#4)

jtsshieh commented on index.d.ts:
```diff
@@ -1831,222 +1862,324 @@ declare namespace Eris {
-    type: string;
+    id: string;
+    type: PermissionType;
     constructor(data: { allow: number; deny: number });
```
id: string, and type: PermissionType are also passed in `data`

```js
    constructor(data) {
        super(data.allow, data.deny);
        this.id = data.id;
        this.type = data.type;
    }
```

bsian03 committed:
Add Error classes

bsian03 committed:
Correct PermissionOverwrite constructor

bsian03 committed:
Missing/correct docs

bsian03 committed:
(╯°□°）╯︵ ┻━┻

jtsshieh committed:
Alphabetize other classes (#5)

bsian03 committed:
Consistency with array types

bsian03 committed:
Correct docs for createMessage

bsian03 committed:
Apply sort class member lint
Class#on doesn't work yet, trying to figure out why

jtsshieh committed:
Capitalise Function types (#6)

bsian03 committed:
Fix linting

bsian03 committed:
TypeScript v4

bsian03 committed:
Fix linting for TS v4

jtsshieh committed:
Add indexers to the Constants (#7)

bsian03 commented on index.d.ts:
```diff
@@ -1008,1044 +1282,986 @@ declare namespace Eris {
+    reset(): void;
+    resetPackets(): void;
+    setVolume(volume: number): void;
+    stop(e: Error, source: NodeJS.WritableStream): void; // Is this correct? There are too many options this could be
```
Need someone to verify this for me, ta

jtsshieh committed:
Correct getVoiceRegions typing (#9)

bsian03 committed:
Deprecate add/removeMessageReaction userID param

alex-taxiera commented on index.d.ts:
```diff
@@ -1,271 +1,829 @@
+// Needed for BrowserWebSocket
+/// <reference lib="dom" />
```
Using this DOM reference will basically inject web types into every project using eris, polluting globals and causing other issues. We need to look for some alternative here.

jtsshieh replied:
is it possible to have a index.dom.ts and have that as the dom typings?

bsian03 replied:
Well we'd need some sort of postinstall/prepare script which prompts the user if they're using a browser, referencing it won't work as it'll import it as full and basically replicate this. I do have some sort of idea which implements this but it's not that great of a workaround since we still need to maintain 2 sets of typings. Caalyx actually suggested to just copy the typings directly from lib.dom.d.ts instead of referencing it so that we wouldn't get global pollution. I'll have a look at this later though

bsian03 replied:
Ended up deciding to just ts-ignore the references instead since the types will be referenced if they are using TypeScript in a browser environment

alex-taxiera commented on index.d.ts:
```diff
@@ -1,271 +1,829 @@
     maxUses?: number;
     temporary?: boolean;
-    unique?: boolean;
+    unique?: boolean; }
```
Closing bracket should be on separate line.

bsian03 replied:
Nice catch, thanks

alex-taxiera commented on index.d.ts:
```diff
@@ -1,271 +1,829 @@
+    owner?: string;
+    prefix?: string | string[];
   }
 
```
Are we keeping a blank line between declarations or not?

bsian03 replied:
That group is actually for the Client, hence why that was in its own group and the other command stuff in a separate group, so yes there should be a blank line here.

alex-taxiera commented on index.d.ts:
```diff
@@ -362,627 +1027,242 @@ declare namespace Eris {
-    description?: string;
-  }
-  interface OldGuild {
+    integrations: any[]; // TODO ????
```
`unknown[]` for now is better.

bsian03 replied:
That didn't actually get picked up by ESLint - should we enable the rule for no explicit any?

alex-taxiera reviewed (changes requested):
Are the types supposed to be grouped into sections? If so then we should label them with comments where the sections begin and end, simply adding a blank line between sections is not enough.

(This will also let us enforce blank line or no blank line between each declaration, as God intended)

That being said it's not a big deal, worst thing here is the triple slash reference to DOM types.

bsian03 committed:
JSDoc consistency

bsian03 committed:
Strict type slack webhook

bsian03 committed:
Use unknown types

bsian03 committed:
ts-ignore DOM references

bsian03 committed:
Match NodeJS version typings

bsian03 committed:
Correct Piper#stop source param type

bsian03 committed:
Mark categories

bsian03 committed:
Document channel ID in edit webhook options

bsian03 committed:
Slack webhook payload not optional

bsian03 committed:
Shard#disconnect options

bsian03 committed:
Have no idea wtf happened here

DonovanDMC committed:
OldGuildChannel.position typings & rawREST jsdoc (#11)

* change OldGuildChannel.position from string to number

* fix typo in rawREST event

eritbh committed:
Add `team` to `OAuthApplicationInfo` (#10)

* add `team` to OAuthApplicationInfo

* break out partial user references for consistency

* move PartialUser to correct category

* Alphabetical order

Co-authored-by: bsian03 <chharry321@gmail.com>

bsian03 committed:
Move deprecation to only PrivateChannel

bsian03 committed:
Lint

abalabahaha committed:
Merge remote-tracking branch 'origin/dev' into typings

abalabahaha reviewed (approved):
This better work
